activerecord/test/cases/encryption/concurrency_test.rb

```diff
@@ -9,18 +9,35 @@ class ActiveRecord::Encryption::ConcurrencyTest < ActiveRecord::EncryptionTestCa
   end
 
   test "models can be encrypted and decrypted in different threads concurrently" do
```

Comment:
This just needlessly creates 800 SQL queries.

Reply:
Agree the current test is overkill, but I don't think a one-shot of aligning two threads to about the same point and then calling a method is usefully equivalent. 

The existing intent is to repeat enough that a wide variety of concurrent execution overlaps are created... I don't have a strong opinion on how valuable that is as an exercise, but if we assume some thread safety issue the existing test would detect, it seems quite unlikely the new one would still catch it -- at best it would show up as a very rare flake.

Reply:
> The existing intent is to repeat enough that a wide variety of concurrent execution overlaps are created

I think there was some concurrency issues back then and this test was implemented to try to catch this in the future. Similar tests (testing for concurrency) can be written for any feature, but I am not sure this should be done and how they are useful. Just a fix + the most basic test would be enough for me.

cc @jorgemanrubia (since you implemented this feature)

Reply:
Hey @matthewd @fatkodima: Active Record encryption supports thread-safe "encryption contexts" when performing encryption operations and it used those in several places. There is a default encryption context which is determined by the attribute being/encrypted or decrypted; also, the support for previous encryption schemes, relies on setting an encryption context when kicking in.

These tests are very important, as they validate that these encryption contexts are thread-safe (as determined by [this property](https://github.com/rails/rails/blob/a199aaedb8e7b54b14874d33345ba8255a3f589b/activerecord/lib/active_record/encryption/contexts.rb#L18)). I added them early in the development of the library when thinking of thread-safety concerns, and I added those tests after realizing of the seriousness of a flaw there. Those tests did reproduce the thread-safety problem I fixed. 

You can reproduce the problem by changing  this line to use `mattr_accessor`:

https://github.com/rails/rails/blob/a199aaedb8e7b54b14874d33345ba8255a3f589b/activerecord/lib/active_record/encryption/contexts.rb#L18

The original test would fail (catching the error), yours would pass.

Reply:
Thank you very much for the response! I am not very familiar with new encryption feature.

But looks like we can still replace `posts = 200.times.collect ..` line with a single `insert_all`, correct?

Reply:
@fatkodima oh yes we can optimize that. We can also reduce that 200 to 100 and the test still catches the problem. So this can replace the `200.times...`:

```ruby
EncryptedPost.insert_all 100.times.collect { |index| { title: "Article #{index} (#{thread_label})", body: "Body #{index} (#{thread_label})" } }
posts = EncryptedPost.last(100)
```

Reply:
Reverted original changes for this test case and rewritten it with `insert_all`.